### PR TITLE
[WIP][DropZone] Allow fixed height on DropZone

### LINKED ISSUE #742: DropZone height is not respected
# Issue summary

The [`<DropZone />`](https://polaris.shopify.com/components/actions/drop-zone) determines its size (height and width) based on width only. This makes certain valid use cases, where the default `DropZone` height isn't appropriate, impossible. For instance, this one:

`DropZone` is configurable to work at various sizes, however. For example:

**Default**

**50 height, 50 width**
screen shot 2018-12-12 at 2 06 50 pm

So far, so good. Until...

**50 height, unspecified width**
screen shot 2018-12-12 at 2 08 19 pm

## Expected behavior

I would expect something more like:

**50 height, unspecified width**
screen shot 2018-12-12 at 2 12 07 pm

## Actual behavior

**50 height, unspecified width**
screen shot 2018-12-12 at 2 08 19 pm

Notice that the DropZone is cut off at the bottom. That is because it actually returns to its default size (shown above).

## Steps to reproduce the problem

1. Go to https://polaris.shopify.com/components/actions/drop-zone
2. Paste this code: `<div style={{height: 50}}><DropZone /></div>`

## Specifications

- Are you using the React components? (Y):
- Polaris version number: latest
- Browser: All
- Device: All
- Operating System: All

### PR DESCRIPTION
**Pull request by elizabethletourneau**
### WHY are these changes introduced?

Resolves #742  

UPDATE: 
When a height is specified the consumer will need to also adjust the horizontal alignment of anything they put within the `DropZone` (including a `FileUpload` component.) 

The `DropZone` determines its size (height and width) based on width only. This makes certain valid use cases, where the default `DropZone` height isn't appropriate, impossible. 

### Important Notes - Needs Discussion ###
This PR was originally merged #908 however there was an issue with putting a flex on `Dropzone_Container` which caused DropZone.FileUpload to centre horizontally however this didn't take into account the fact that DropZone could have other children.

What is up for discussion is that when a fixed height is given to the DropZone should it be up to the consumer to style DropZone's children or should we have two conditions, one that takes into account if FileUpload is the only child (therefore center it ) and the other being do not center FileUpload when there are multiple children.

There is also the fix Dan R. proposes [here](https://github.com/Shopify/polaris-react/pull/1127) which adds a prop that center-aligns children and defaults to true.

### WHAT is this pull request doing?

This pull request allows the consumer to set a fixed height on a wrapping parent element so that the `DropZone` takes its parent's height if one is provided.

Current Behaviour:  (wrapped with a div given a height of 50px)
image

New Behaviour: (wrapped with a div given a height of 50px)
image

image

image

(Note: there are in between size variations (ie. small width, medium height etc.) The `FileUpload` will always render the smaller measurement of width or height.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)
<details>
<summary>Example of All Dropzone Possibilities </summary>

```jsx
import * as React from 'react';
import {
  Page,
  DropZone,
  Stack,
  Thumbnail,
  Caption,
  Heading,
  Subheading,
} from '../src';

interface State {
  files: object;
}

export default class Playground extends React.Component<{}, State> {
  state = {
    files: [],
  };

  render() {
    const {files} = this.state;
    const validImageTypes = ['image/gif', 'image/jpeg', 'image/png'];

    const fileUpload = !files.length && <DropZone.FileUpload />;
    const uploadedFiles = files.length > 0 && (
      <Stack vertical>
        {files.map((file, index) => (
          <Stack alignment="center" key={index}>
            <Thumbnail
              size="small"
              alt={file.name}
              source={
                validImageTypes.indexOf(file.type) > 0
                  ? window.URL.createObjectURL(file)
                  : 'https://cdn.shopify.com/s/files/1/0757/9955/files/New_Post.png?12678548500147524304'
              }
            />
            <div>
              {file.name} <Caption>{file.size} bytes</Caption>
            </div>
          </Stack>
        ))}
      </Stack>
    );

    return (
      <Page title="Dropzones">
        <Heading> Small Height : Small Width </Heading>
        <div style={{width: 50, height: 50}}>
          <DropZone
            onDrop={(files, acceptedFiles, rejectedFiles) => {
              this.setState({
                files: [...files, ...acceptedFiles],
              });
            }}
          >
            {uploadedFiles}
            {fileUpload}
          </DropZone>
        </div>
        <Heading> Medium Height : Medium Width </Heading>
        <div style={{width: 150, height: 150}}>
          <DropZone
            onDrop={(files, acceptedFiles, rejectedFiles) => {
              this.setState({
                files: [...files, ...acceptedFiles],
              });
            }}
          >
            {uploadedFiles}
            {fileUpload}
          </DropZone>
        </div>
        <Heading> Large Height : Large Width </Heading>
        <div style={{width: 250, height: 250}}>
          <DropZone
            onDrop={(files, acceptedFiles, rejectedFiles) => {
              this.setState({
                files: [...files, ...acceptedFiles],
              });
            }}
          >
            {uploadedFiles}
            {fileUpload}
          </DropZone>
        </div>
        <Heading> ExtraLarge Height : ExtraLarge Width </Heading>
        <div style={{width: 500, height: 500}}>
          <DropZone
            onDrop={(files, acceptedFiles, rejectedFiles) => {
              this.setState({
                files: [...files, ...acceptedFiles],
              });
            }}
          >
            {uploadedFiles}
            {fileUpload}
          </DropZone>
        </div>
        <Heading> Default Height : Default Width </Heading>
        <Subheading>mobile and full width</Subheading>
        <DropZone
          onDrop={(files, acceptedFiles, rejectedFiles) => {
            this.setState({
              files: [...files, ...acceptedFiles],
            });
          }}
        >
          {uploadedFiles}
          {fileUpload}
        </DropZone>
        <Heading> ExtraLarge Height : Small Width </Heading>
        <div style={{width: 50, height: 500}}>
          <DropZone
            onDrop={(files, acceptedFiles, rejectedFiles) => {
              this.setState({
                files: [...files, ...acceptedFiles],
              });
            }}
          >
            {uploadedFiles}
            {fileUpload}
          </DropZone>
        </div>
        <Heading> Small Height : ExtraLarge Width </Heading>
        <div style={{width: 500, height: 50}}>
          <DropZone
            onDrop={(files, acceptedFiles, rejectedFiles) => {
              this.setState({
                files: [...files, ...acceptedFiles],
              });
            }}
          >
            {uploadedFiles}
            {fileUpload}
          </DropZone>
        </div>
      </Page>
    );
  }
}

```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)

### CODE REVIEW

**Comment by elizabethletourneau** [comment]
@dleroux @ry5n @danrosenthal I'm tagging you all to discuss the options that this PR produces and to also make sure I didn't miss anything in my notes above about the options. Happy to also keep looking into solutions for this in the meantime! Thank you.

**Comment by nairim** [comment]
Hi @elizabethletourneau! @StephPoce talked about this issue today. I just want to let you know that I'll take a look at that from a design perspective. Let me organize my next few days and all the catch up after vacation.

**Comment by nairim** [comment]
So I took a look at the related issues, and @elizabethletourneau helped me to understand a bit more of the problem (thank you, Liz! <3)

Taking in consideration the fact that `Boolean props should often be false by default` (unless there’s a legitimate exception) and, paraphrasing @danrosenthal on issue #1127, “its not typically desirable to have parents lay out their children” I believe is safe to accept that it should be up to the consumer to style drop zone’s children if they want to customize the height of the component.

I thought that maybe there’s an opportunity here to add some guidelines to the style guide in how to do centralize it horizontally once you customize the height of the drop zone.

Unrelated question: I was playing with the props and tried the `Disabled` one. Specifically, when we have the option of having a “Add file” button, should the button appear on the disabled state if the disabled prop is set to `true`?

File: src/components/DropZone/DropZone.tsx [hunk 1]
```diff
@@ -383,6 +398,27 @@ export class DropZone extends React.Component<CombinedProps, State> {
...
+    const getSize = size === 'height' ? 'height' : 'width';
+    if (node.current) {
+      const wrapper = node.current.getBoundingClientRect()[getSize];
+      if (wrapper < Size.Small) {
+        wrapperSize = Size.Small;
+      } else if (wrapper < Size.Medium) {
+        wrapperSize = Size.Medium;
+      } else if (wrapper < Size.Large) {
+        wrapperSize = Size.Large;
+      } else {
```
[the remark this answers is not in the document]
Comment (elizabethletourneau) [reply src/components/DropZone/DropZone.tsx]: 🔥 🔥 🔥

File: UNRELEASED.md [hunk 2]
```diff
@@ -23,6 +23,7 @@ Use [the changelog guidelines](https://git.io/polaris-changelog-guidelines) to f
 - Fixed selected state for date picker in windows high contrast mode ([#1342](https://github.com/Shopify/polaris-react/pull/1342))
 - Added background into media query for Microsoft high contrast to fix skeleton accessibility. ([#1341](https://github.com/Shopify/polaris-react/pull/1341))
 - Fixed the position calculation of the `PositionedOverlay` component after scroll. ([#1382](https://github.com/Shopify/polaris-react/pull/1382))
+  Constrain DropZone height based on inherited wrapper height ([#1138](https://github.com/Shopify/polaris-react/pull/1138))
```
Comment (AndrewMusgrave) [inline UNRELEASED.md]:
```suggestion
- Constrain DropZone height based on inherited wrapper height ([#1138](https://github.com/Shopify/polaris-react/pull/1138))
```

**Comment by danrosenthal** [comment]
Removing my review request, as I can't commit to reviewing at this time

**Comment by chloerice** [comment]
Hey @AndrewMusgrave, should I rebase and merge this, or are we still looking for design input first?

**Comment by mirualves** [comment]
I think we're fine on the design, @chloerice !

File: src/components/DropZone/DropZone.tsx [hunk 3]
```diff
@@ -383,6 +398,27 @@ export class DropZone extends React.Component<CombinedProps, State> {
     this.fileInputNode.current.click();
   };
 
+  private setWrapperSize = (
+    size: string,
+    node: React.RefObject<HTMLDivElement>,
+  ) => {
+    let wrapperSize = Size.ExtraLarge;
+    const getSize = size === 'height' ? 'height' : 'width';
```
[the remark this answers is not in the document]
Comment (chloerice) [reply src/components/DropZone/DropZone.tsx]: 💯

**Comment by dleroux** [comment]
@chloerice I just rebased and tests are failing. Any idea what changed? I can dig into it but just wondering if you knew before spending too much time on this.

**Comment by chloerice** [comment]
> @chloerice I just rebased and tests are failing. Any idea what changed? I can dig into it but just wondering if you knew before spending too much time on this.

Tests started failing after my rebase as well. I haven't been successful in finding what's going on since the rebase, was digging into the higher priority issue. I'll be looking at this again Wednesday morning, but feel free to poke at it in the meantime!

**Comment by dleroux** [comment]
I've rebased this and fixed tests. 

@mirualves you're good with the Percy changes?

@dpersing  note the a11y issues. What's the recommendation here?

**Comment by mirualves** [comment]
@dleroux  Let's investigate why it's affecting the spacing between the illustration, button and hint text. Ideally, it should have the same spacing as before.

**Comment by dleroux** [comment]
I just noticed that when there's a label and it's hidden small renders the large height:

Screen Shot 2019-07-25 at 12 57 42 PM

**Comment by dleroux** [comment]
I checked the spacing @mirualves and the Stack is behaving correctly. The new dimensions are what is causing it so I think it's fine.

**Comment by dleroux** [comment]
Now that label is always required something is broken with the height.

**Comment by dpersing** [comment]
> note the a11y issues. What's the recommendation here?

@dleroux Are you referencing automated tests (which are passing now, it looks), or the a11y issues logged against the component?

**Comment by dleroux** [comment]
Yes @dpersing. Sorry I figured it out. The accessibility fix (always having a label) is what breaks the visual layout right now.

**Comment by danrosenthal** [comment]
This PR has been open for about 6 months, I'm thinking we ought to close it until someone can commit to this change and see it through.

**Comment by dleroux** [comment]
You see an issue with leaving it open? I'm ok with closing this but I just worry it might be completely forgotten.

**Comment by dpersing** [comment]
@danrosenthal @daniedleroux Closing the PR should be okay if the associated issue remains open?

**Pull request closed without merging**